### PR TITLE
Board generator v2

### PR DESCRIPTION
**Pull request by RandyBrown1965**
This second version of the board generator uses a random walks method.

The head_target_erode method of generating wires is still in-progress because it creates extraneous loops.

### CODE REVIEW

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+from constants import SOURCE as WIRE
+# Also available to import from constants OBSTACLE, NOOP, LEFT, LEFT, UP, RIGHT, DOWN
+from dataclasses import dataclass
+import numpy as np
+from copy import deepcopy
+import random
+from typing import List
+
+@dataclass
+class Position():
```
Comment (mwolinska) [inline jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py:11]: unnecessary brackets

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+    Heads are encoded starting from 4 in multiples of 3: 4, 7, 10, ...
+    Targets are encoded starting from 3 in multiples of 3: 3, 6, 9, ...
+    Wiring routes connecting the head/target pairs are encoded starting
+	    at 2 in multiples of 3: 2, 5, 8, ...
+
+    Args:
+        x_dim, ydim (int, int) : Dimensions of the board.
+
+    """
+    def __init__(self, x_dim: int, y_dim: int) -> None:
```
Comment (mwolinska) [inline jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py:31]:
```suggestion
    def __init__(self, x_dim: int, y_dim: int): 
```

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py (hunk 1 above)
Comment (RandyBrown1965) [reply jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py]: Removed.

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+    def add_wire_start_distance_directions(self) -> None:
+        """
+        Define a wire and add it to the layout of the board.
+
+        The wire consists of a head cell, a target cell, and any connecting wire.
+        This implementation generates the wire using a random starting point,
+        then proceeds the given number of steps in the two principle directions.
+        """
+        invalid_head = True 
+        while invalid_head == True:
```
Comment (mwolinska) [inline jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py:76]:
```suggestion
        while invalid_head:
```

**Review by mwolinska — COMMENTED** [review]
Review docstrings before merging

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+        self.dim = Position(x_dim, y_dim)
+        self.num_wires = 0
+
+    def get_random_head(self) -> Position:
+        # Return a random 2D position, a starting point in the array
+        x_dim = random.randint(0, self.dim.x-1) 
+        y_dim = random.randint(0, self.dim.y-1) 
+        return Position(x_dim, y_dim)
+
+    def get_wiring_directions(self, head: Position) -> (Position, Position):
```
[the remark this answers is not in the document]
Comment (RandyBrown1965) [reply jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py]: Done

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    """
+    layout_out = []
+    for row_in in layout_in:
+        # Zero out any element that is a connecting wire
+        row_out = [i * int(i%3 != WIRE) for i in row_in]
+        layout_out.append(row_out)
+    return np.array(layout_out)
+
+
+def board_generator(x_dim: int, y_dim: int, target_wires: int = None) -> (np.ndarray, np.ndarray, int):
```
[the remark this answers is not in the document]
Comment (RandyBrown1965) [reply jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py:397]: Sure.  We can change this if we coordinate with the others.

File: jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+           Returns:
+               bool : True if the cell is an extraneous corner that can be removed,
+                    e.g. it has two adjacent empty cells, and the cell in the opposite corner is full.
+        """
+        # Initialize variables.
+        (x,y) = cell
+        upper_empty, left_empty, bottom_empty, right_empty = False, False, False, False
+        botright_full, botleft_full, upright_full, upleft_full = False, False, False, False
+        # Check for empty adjacent cells
+        if (x-1, y) not in connectible_list:
```
[the remark this answers is not in the document]
Comment (RandyBrown1965) [reply jumanji/environments/combinatorial/routing/board_generator_v2.0.0_rb.py]: Done

**Pull request merged**